tap_salesforce/client.py

```diff
@@ -0,0 +1,356 @@
+
+        headers = {
+            "Authorization": "Bearer {}".format(self._access_token),
+            "Sforce-Query-Options": "batchSize=500",
+        }
+
+        url = f"{self._instance_url}{path}"
+        resp = self.session.request(
+            method, url, headers=headers, params=params, data=data
+        )
+
+        resp.raise_for_status()
+
+        self._metrics_http_requests += 1
+        self._check_rest_quota_usage(resp.headers)
+
+        return resp
+
+    def _login(self):
+        if self.is_sandbox:
+            login_url = "https://test.salesforce.com/services/oauth2/token"
+        else:
+            login_url = "https://login.salesforce.com/services/oauth2/token"
+
+        data = {
+            "grant_type": "refresh_token",
+            "client_id": self.client_id,
+            "client_secret": self.client_secret,
+            "refresh_token": self.refresh_token,
+        }
+
+        LOGGER.info("Attempting login via OAuth2")
+
+        try:
+            resp = self.session.post(
+                login_url,
+                headers={"Content-Type": "application/x-www-form-urlencoded"},
+                data=data,
+            )
+
+            resp.raise_for_status()
+
+            LOGGER.info("OAuth2 login successful")
+            auth = resp.json()
+
+            self._access_token = auth["access_token"]
+            self._instance_url = auth["instance_url"]
+
+            self._token_expiration_time = datetime.utcnow() + timedelta(
+                seconds=self._REFRESH_TOKEN_EXPIRATION_PERIOD
+            )
+        except requests.exceptions.HTTPError as req_ex:
+            response_text = None
+            if req_ex.response:
+                response_text = req_ex.response.text
+                LOGGER.exception(response_text or str(req_ex))
+                if req_ex.response.status_code == 403:
+                    raise TapSalesforceOauthException(
+                        f"invalid oauth2 credentials: {req_ex.response.text}"
+                    )
+            raise TapSalesforceOauthException(
+                "failed to refresh or login using oauth2 credentials"
+            )
+
+    def _check_rest_quota_usage(self, headers):
+        match = re.search(r"^api-usage=(\d+)/(\d+)$", headers.get("Sforce-Limit-Info"))
+
+        if match is None:
+            return
+
+        used, total = map(int, match.groups())
+
+        LOGGER.info(
+            f"Used {used / total * 100:.2f}% of daily REST API quota",
+        )
+
+        used_percent = (used / total) * 100.0
+
+        max_requests = int((self.quota_percent_per_run * total) / 100)
+
+        if used_percent > self.quota_percent_total:
+            total_message = (
+                "Salesforce has reported {}/{} ({:3.2f}%) total REST quota "
+                + "used across all Salesforce Applications. Terminating "
+                + "replication to not continue past configured percentage "
+                + "of {}% total quota."
```

Comment:
I really like the percentage way showing used quota. Much more intuitive.

Reply:
yeah, the old thing was driving me a little bit nuts.. Still very noisy code, but I see no way of reducing it without keeping a state of the previous percentage, and then only writing it if it changed..